### DISCUSSION TITLE
Presentation on llama.cpp on 25.07.2023 at karlsruhe.ai

### DISCUSSION THREAD

**Discussion by JohannesGaessler** [ggml-org/llama.cpp#2281] [Show and tell]
I will hold a presentation on llama.cpp (mostly the things that I'm doing myself) on 25.07.2023 at [karlsruhe.ai](https://www.meetup.com/karlsruhe-ai/events/294737034). I am uploading the corresponding slides [here](https://github.com/JohannesGaessler/presentations/tree/master/2023_07_25_llama.cpp_GPU_Acceleration). They are still missing some things that I'll measure over the weekend but the slides should be mostly complete in terms of content. I'm currently using a low-effort stable diffusion gen of a llama on a motorcycle for the cover image; if someone gens me something better I'll use it, otherwise I'll probably look into locally installing Stable Diffusion again after not having used it for several months.

<details><summary>The current cover image</summary>

![A llama riding an NVIDIA gaming motorcycle at night, a cool alpaca, starry sky, award-winning photograph](https://github.com/ggerganov/llama.cpp/assets/18492268/2c5381c4-c731-46d1-a2f4-14b63cf8ddeb)

</details>

@ggerganov please check whether you agree with the way I'm representing the llama.cpp project.

@ikawrakow please check whether you agree with my descriptions of k-quants.

**Comment by ggerganov** [comment]
@JohannesGaessler 

Thanks for sharing the presentation! The slides look fine to me - I like the "Open-source is a positive sum game" quote :)
There are many interesting details around the CUDA implementation that you can delve into depending on your time slot.
You can maybe add an overview of the LLaMA computation graph, though it might be too much information for an oral presentation:

![llama dot](https://github.com/ggerganov/llama.cpp/assets/1991296/9ecc8283-4cc0-4d5e-b387-a6e82b2888a7)

Definitely let us know how it goes!

**Comment by JohannesGaessler** [comment]
I did some LLaMA vs. LLaMA 2 measurements for the presentation where I calculated perplexity as a function of context length on a source code dump of [kafe2](https://github.com/PhiLFitters/kafe2), another project that I'm working on. I chose this over Wikitext because I wanted text with a large amount of long-distance relations. These are the results:

![060_llama_vs_llama_2](https://github.com/ggerganov/llama.cpp/assets/18492268/869f06a7-a644-476c-8706-10b9ffdfbc5c)

As it turns out the perplexity of LLaMA 2 7b at 4096 context is actually better than the perplexity of LLaMA 13b at 2048 context.

**Reply by xx205** [reply]
Current llama.cpp code has an inconsistent rms_norm_eps for LLaMA-2, hence the estimated PPLs are higher than what they could be. https://github.com/ggerganov/llama.cpp/issues/2373

**Reply by elix1er** [reply]
can we expect quality or speed inprovement

**Reply by xx205** [reply]
quality

**Reply by JohannesGaessler** [reply]
Thank you for notifying me. I reran the calculations using the modified eps and the perplexity is noticeably better:

![060_llama_vs_llama_2](https://github.com/ggerganov/llama.cpp/assets/18492268/32e3fc27-0ee8-44f9-8907-82a46ab07c20)

**Comment by JohannesGaessler** [comment]
The presentation was well-received, ~20 people were in the audience, mostly other KIT members. I had some interesting conversations but unfortunately I didn't meet a language model researcher that I could collaborate with specifically. I uploaded the slides that I used as well as the SVG and Python files that I used for creating the graphics in case someone wants to recycle them for documentation purposes.

**Reply by ggerganov** [reply]
Thanks!

The prompt processing speed for `llama.cpp` on slide 14 seems a bit off. I can observe ~1700 t/s with `perplexity` tool by using `-t 1 -ngl 1000` for 7B models on RTX 4080. Is it really that much faster than RTX 3090?

**Reply by JohannesGaessler** [reply]
It depends on which numbers you use for comparison. The "prompt eval time" and "eval time" overestimate llama.cpp performance if they are fast compared to e.g. building the ggml graph. The prompt processing t/s reported by llama.cpp on my RTX 3090 are 1325, the number in the presentation is based on timing the total process runtime via the `time` command.
